crates/biome_js_analyze/src/lint/nursery/no_useless_length_check.rs

```diff
@@ -0,0 +1,366 @@
+use std::collections::{HashMap, HashSet};
+
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+    RuleSource,
+};
+use biome_console::markup;
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsLiteralExpression, AnyJsMemberExpression, JsBinaryExpression,
+    JsBinaryOperator, JsLogicalExpression, JsParenthesizedExpression, JsSyntaxKind, T,
+};
+use biome_rowan::{AstNode, BatchMutationExt, TextRange};
+
+use crate::JsRuleAction;
+
+declare_lint_rule! {
+    /// Disallow unnecessary length checks within logical expressions.
+    ///
+    /// - `Array#some()` returns `false` for an empty array. There is no need to check if the array is not empty.
+    /// - `Array#every()` returns `true` for an empty array. There is no need to check if the array is empty.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// if (array.length === 0 || array.every(Boolean));
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// if (array.length !== 0 && array.some(Boolean));
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// if (array.length > 0 && array.some(Boolean));
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const isAllTrulyOrEmpty = array.length === 0 || array.every(Boolean);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// if (array.every(Boolean));
+    /// ```
+    ///
+    /// ```js
+    /// if (array.some(Boolean));
+    /// ```
+    ///
+    /// ```js
+    /// const isAllTrulyOrEmpty = array.every(Boolean);
+    /// ```
+    ///
+    /// ```js
+    /// const isAllTrulyOrEmpty = array.every(Boolean);
+    /// ```
+    ///
+    /// ```js
+    /// if (array.length === 0 || anotherCheck() || array.every(Boolean));
+    /// ```
+    ///
+    /// ```js
+    /// const isNonEmptyAllTrulyArray = array.length > 0 && array.every(Boolean);
+    /// ```
+    ///
+    /// ```js
+    /// const isEmptyArrayOrAllTruly = array.length === 0 || array.some(Boolean);
+    /// ```
+    ///
+    pub NoUselessLengthCheck {
+        version: "next",
+        name: "noUselessLengthCheck",
+        language: "js",
+        recommended: false,
+        sources: &[RuleSource::EslintUnicorn("no-useless-length-check")],
+        fix_kind: FixKind::Unsafe,
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum ErrorType {
+    UselessLengthCheckWithSome,
+    UselessLengthCheckWithEvery,
+}
```

Comment:
```suggestion
/// Tracks which function was used
#[derive(Clone, Debug)]
pub enum FunctionKind {
    /// `Array.some()` was used 
    Some,
    /// `Array.every()` was used
    Every,
}
```

Nit: I think it's implied we're looking for a violation, so using the word "error" is superfluous.